pkg/serviceaccounttoken/secret.go

```diff
@@ -25,74 +26,102 @@ const (
 // secretLister is an abstraction over any kind of secret lister.
 // The caller can use any cache or client it has available, whether that is from norman, wrangler, or client-go,
 // as long as it can wrap it in a simplified lambda with this signature.
-type secretLister func(namespace string, selector labels.Selector) ([]*v1.Secret, error)
+type secretLister func(namespace string, selector labels.Selector) ([]*corev1.Secret, error)
+
+var lockMap sync.Map
+
+func getLock(key string) *sync.Mutex {
+	actual, _ := lockMap.LoadOrStore(key, &sync.Mutex{})
+	return actual.(*sync.Mutex)
+}
 
 // EnsureSecretForServiceAccount gets or creates a service account token Secret for the provided Service Account.
-// For k8s <1.24, the secret is automatically generated for the service account. For >=1.24, we need to generate it explicitly.
-func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrollers.SecretCache, clientSet kubernetes.Interface, sa *v1.ServiceAccount) (*v1.Secret, error) {
+//
+// If lockPrefix is provided, this is used as a prefix for the lock to provide
+// context for concurrent requests. No concurrent creates will be allowed for a
+// service account.
+//
+// For example, if this is "cluster-1" and the ServiceAccount is
+// "default:test-sa" subsequent requests to create a ServiceAccount with the
+// same name in "cluster-1" will be blocked until the first request has
+// completed.
+//
+// Note: This mutex is per-replica, currently there's no attempt to co-ordinate
+// across replicas.
+func EnsureSecretForServiceAccount(ctx context.Context, secretsCache corecontrollers.SecretCache, clientSet kubernetes.Interface, sa *corev1.ServiceAccount, lockPrefix string) (*corev1.Secret, error) {
```

Comment:
Nit: Might be good to specify what "LockPrefix" should be since there seems to be a convention for it (empty if in the same cluster, `utils.FormatPrefix(cluster.Name)` if it's not).